core/solver/batch_gmres.cpp

```diff
@@ -0,0 +1,143 @@
+
+
+#include <ginkgo/core/matrix/batch_dense.hpp>
+
+
+#include "core/solver/batch_gmres_kernels.hpp"
+
+
+namespace gko {
+namespace solver {
+namespace batch_gmres {
+
+
+GKO_REGISTER_OPERATION(apply, batch_gmres::apply);
+
+
+}  // namespace batch_gmres
+
+
+template <typename ValueType>
+std::unique_ptr<BatchLinOp> BatchGmres<ValueType>::transpose() const
+{
+    return build()
+        .with_preconditioner(parameters_.preconditioner)
+        .with_max_iterations(parameters_.max_iterations)
+        .with_rel_residual_tol(parameters_.rel_residual_tol)
+        .with_abs_residual_tol(parameters_.abs_residual_tol)
+        .with_restart(parameters_.restart)
+        .with_tolerance_type(parameters_.tolerance_type)
+        .on(this->get_executor())
+        ->generate(share(
+            as<BatchTransposable>(this->get_system_matrix())->transpose()));
+}
+
+
+template <typename ValueType>
+std::unique_ptr<BatchLinOp> BatchGmres<ValueType>::conj_transpose() const
+{
+    return build()
+        .with_preconditioner(parameters_.preconditioner)
+        .with_max_iterations(parameters_.max_iterations)
+        .with_rel_residual_tol(parameters_.rel_residual_tol)
+        .with_abs_residual_tol(parameters_.abs_residual_tol)
+        .with_restart(parameters_.restart)
+        .with_tolerance_type(parameters_.tolerance_type)
+        .on(this->get_executor())
+        ->generate(share(as<BatchTransposable>(this->get_system_matrix())
+                             ->conj_transpose()));
+}
+
+
+template <typename ValueType>
+void BatchGmres<ValueType>::apply_impl(const BatchLinOp *b, BatchLinOp *x) const
+{
+    using Vector = matrix::BatchDense<ValueType>;
+    using real_type = remove_complex<ValueType>;
+
+    auto exec = this->get_executor();
+    auto dense_b = as<const Vector>(b);
+    auto dense_x = as<Vector>(x);
+    const kernels::batch_gmres::BatchGmresOptions<remove_complex<ValueType>>
+        opts{parameters_.preconditioner,   parameters_.max_iterations,
+             parameters_.rel_residual_tol, parameters_.abs_residual_tol,
+             parameters_.restart,          parameters_.tolerance_type};
+
+    log::BatchLogData<ValueType> logdata;
+
+    // allocate logging arrays assuming uniform size batch
+    // GKO_ASSERT(dense_b->stores_equal_sizes());
+
+    const size_type num_rhs = dense_b->get_size().at(0)[1];
+    const size_type num_batches = dense_b->get_num_batch_entries();
+    batch_dim<> sizes(num_batches, dim<2>{1, num_rhs});
+
+    logdata.res_norms =
+        matrix::BatchDense<real_type>::create(this->get_executor(), sizes);
+    logdata.iter_counts.set_executor(this->get_executor());
+    logdata.iter_counts.resize_and_reset(num_rhs * num_batches);
+
+    exec->run(batch_gmres::make_apply(
+        opts, system_matrix_.get(), this->get_left_scaling_vector(),
+        this->get_right_scaling_vector(), dense_b, dense_x, logdata));
+
+    this->template log<log::Logger::batch_solver_completed>(
```

Comment:
Shouldn't there also be a `batch_solver_started` log ?

Reply:
Right now, there's actually no such log function, because it only logs the iteration counts and residual norms. For the timings, I guess it's the BatchLinOp `apply` logging that's being used?